pkg/operator/ceph/object/rgw.go

```diff
@@ -91,55 +104,119 @@ func (c *clusterConfig) createOrUpdate(update bool) error {
 		}
-		err = k8sutil.DeleteDaemonset(c.context.Clientset, c.store.Namespace, c.instanceName())
-		if err != nil {
-			logger.Warning(err.Error())
+		daemonsetNum := len(daemons.Items)
+		if daemonsetNum > 0 {
+			for i, d := range daemons.Items {
+				// Delete any existing daemonset
+				if err := k8sutil.DeleteDaemonset(c.context.Clientset, c.store.Namespace, d.Name); err != nil {
+					return fmt.Errorf("error during deletion of daemonset %s resource: %+v", d.Name, err)
+				}
+
+				// Create a new deployment accordingly
+				daemonLetterID := k8sutil.IndexToName(i)
+				// Each rgw is id'ed by <store_name>-<letterID>
+				daemonName := fmt.Sprintf("%s-%s", c.store.Name, daemonLetterID)
+				// resource name is rook-ceph-rgw-<store_name>-<daemon_name>
+				resourceName := fmt.Sprintf("%s-%s-%s", AppName, c.store.Name, daemonLetterID)
+				rgwConfig := &rgwConfig{
+					ResourceName: resourceName,
+					DaemonID:     daemonName,
+				}
+
+				deployment, err := c.startDeployment(rgwConfig)
+				if err != nil {
+					return err
+				}
+
+				resourceControllerOwnerRef := &metav1.OwnerReference{
+					UID:        deployment.UID,
+					APIVersion: "v1",
+					Kind:       "deployment",
+					Name:       rgwConfig.ResourceName,
+				}
+
+				// Generate the keyring after starting the replication controller so that the keyring may use
+				// the controller as its owner reference; the keyring is deleted with the controller
+				err = c.generateKeyring(resourceControllerOwnerRef)
+				if err != nil {
+					return fmt.Errorf("failed to create rgw keyring. %+v", err)
+				}
+
+				// Generate the mime.types file after the rep. controller as well for the same reason as keyring
+				if err := c.generateMimeTypes(resourceControllerOwnerRef); err != nil {
+					return fmt.Errorf("failed to generate the rgw mime.types config. %+v", err)
+				}
+
+			}
+			// Remove the old rgw key once all the old rgws (daemonsets) have been replaced
+			err = client.AuthDelete(c.context, c.store.Namespace, "client.radosgw.gateway")
+			if err != nil {
+				return err
+			}
+		} else {
+			logger.Infof("%d rgw daemonset(s) found for store %s in namespace %s", daemonsetNum, c.store.Name, c.store.Namespace)
+			// it looks like there is no pre-existing config so let's set it to false always
+			c.store.Spec.Gateway.AllNodes = false
+			// Now we fail asking users to update their CR
+			return fmt.Errorf("deploying rgw on all the nodes is not supported anymore, please use 'instances' instead")
 		}
-	}
-
-	// start the deployment or daemonset
-	var uid types.UID
-	var controllerType string
-	if c.store.Spec.Gateway.AllNodes {
-		daemonSet, err := c.startDaemonset()
-		if err != nil {
-			return err
-		}
-		uid = daemonSet.UID
-		controllerType = "daemonset"
 	} else {
-		deployment, err := c.startDeployment()
-		if err != nil {
-			return err
+		// start a new deployment
+		for i := 0; i < int(c.store.Spec.Gateway.Instances); i++ {
```

Comment:
To summarize previous comments, what about the following approach? 
- For 1 to `instances` rgw deployments
   - Create the keyring and new deployment
- If there are more than n deployments, remove the extras (if they reduced the `instances` in the CR)
- If allNodes is true, delete the legacy daemonset (ignore error if it doesn't exist). If they were using this option, perhaps we set `instances` temporarily in memory to the number of nodes so they still have enough daemons. Or perhaps we just document it as a breaking change since we don't know that anyone is using it.
- If allNodes is false, delete the legacy deployment (ignore error if it doesn't exist)